File: ios/Namu/Native/GGUFCheck.swift
import Foundation

/// Bounded structural GGUF check (DL-010, contract §6.4 step 3).
/// Runs only on files whose SHA-256 already matched the signed descriptor and
/// never uses the inference runtime's parser. Streams the metadata section
/// through a small buffer; tensor data is never touched and the file is never
/// loaded into memory. Mirrors model-release/dev/gguf-header.mjs.
enum GGUFCheck {
  struct Failure: Error, Equatable {
    let reason: String
  }

  static let maxKeyValues: UInt64 = 4096
  static let maxTensors: UInt64 = 65_536
  static let maxKeyLength: UInt64 = 1 << 20
  static let maxArchitectureLength: UInt64 = 256
  /// The metadata walk may not reach past this offset.
  static let headerWindow: UInt64 = 64 * 1024 * 1024

  private static let scalarSizes: [UInt32: UInt64] = [
    0: 1, 1: 1, 2: 2, 3: 2, 4: 4, 5: 4, 6: 4, 7: 1, 10: 8, 11: 8, 12: 8,
  ]
  private static let typeString: UInt32 = 8
  private static let typeArray: UInt32 = 9

  /// Returns `general.architecture`. Any bound violation throws `Failure`.
  static func readArchitecture(at url: URL) throws -> String {
    let handle: FileHandle
    do {
      handle = try FileHandle(forReadingFrom: url)
    } catch {
      throw Failure(reason: "open failed")
    }
    defer { try? handle.close() }
    let size: UInt64
    do {
      size = try handle.seekToEnd()
      try handle.seek(toOffset: 0)
    } catch {
      throw Failure(reason: "seek failed")
    }
    var reader = BoundedReader(handle: handle, limit: min(size, headerWindow))

    let magic = try reader.read(4)
    guard magic == Data("GGUF".utf8) else { throw Failure(reason: "bad magic") }
    let version = try reader.u32()
    guard version == 2 || version == 3 else { throw Failure(reason: "unsupported version") }
    let tensors = try reader.u64()
    let keyValues = try reader.u64()
    guard tensors <= maxTensors, keyValues <= maxKeyValues else { throw Failure(reason: "counts out of bounds") }

    for _ in 0..<keyValues {
      let key = try reader.string(limit: maxKeyLength)
      let type = try reader.u32()
      if key == "general.architecture" {
        guard type == typeString else { throw Failure(reason: "architecture is not a string") }
        return try reader.string(limit: maxArchitectureLength)
      }
      try skipValue(&reader, type: type, depth: 0)
    }
    throw Failure(reason: "general.architecture missing")
  }

  private static func skipValue(_ reader: inout BoundedReader, type: UInt32, depth: Int) throws {
    if let size = scalarSizes[type] {
      try reader.skip(size)
    } else if type == typeString {
      try reader.skip(try reader.u64())
    } else if type == typeArray {
      guard depth <= 1 else { throw Failure(reason: "nested arrays not allowed") }
      let inner = try reader.u32()
      let count = try reader.u64()
      if let size = scalarSizes[inner] {
        let (bytes, overflow) = size.multipliedReportingOverflow(by: count)
        guard !overflow else { throw Failure(reason: "array too large") }
        try reader.skip(bytes)
      } else {
        // Every element consumes at least four bytes, so the window bounds
        // this loop even for a hostile count.
        guard count <= headerWindow else { throw Failure(reason: "array too large") }
        for _ in 0..<count { try skipValue(&reader, type: inner, depth: depth + 1) }
      }
    } else {
      throw Failure(reason: "unknown value type")
    }
  }

  /// Sequential reader over the first `limit` bytes with a 64 KiB buffer.
  private struct BoundedReader {
    let handle: FileHandle
    let limit: UInt64
    private var buffer = Data()
    private var bufferStart: UInt64 = 0 // file offset of buffer[0]
    private var position: UInt64 = 0 // absolute file offset of the cursor
    private static let chunk = 64 * 1024

    init(handle: FileHandle, limit: UInt64) {
      self.handle = handle
      self.limit = limit
    }

    private func need(_ count: UInt64) throws {
      let (end, overflow) = position.addingReportingOverflow(count)
      guard !overflow, end <= limit else { throw Failure(reason: "metadata exceeds header window") }
    }

    mutating func skip(_ count: UInt64) throws {
      try need(count)
      position += count
    }

    mutating func read(_ count: Int) throws -> Data {
      try need(UInt64(count))
      var out = Data(capacity: count)
      while out.count < count {
        let bufferEnd = bufferStart + UInt64(buffer.count)
        if position < bufferStart || position >= bufferEnd {
          do {
            try handle.seek(toOffset: position)
            let wanted = Int(min(UInt64(max(BoundedReader.chunk, count - out.count)), limit - position))
            buffer = try handle.read(upToCount: wanted) ?? Data()
          } catch {
            throw Failure(reason: "read failed")
          }
          bufferStart = position
          guard !buffer.isEmpty else { throw Failure(reason: "truncated metadata") }
        }
        let offset = Int(position - bufferStart)
        let take = min(count - out.count, buffer.count - offset)
        out.append(buffer.subdata(in: (buffer.startIndex + offset)..<(buffer.startIndex + offset + take)))
        position += UInt64(take)
      }
      return out
    }

    mutating func u32() throws -> UInt32 {
      let data = try read(4)
      return data.enumerated().reduce(UInt32(0)) { $0 | UInt32($1.element) << (8 * UInt32($1.offset)) }
    }

    mutating func u64() throws -> UInt64 {
      let data = try read(8)
      return data.enumerated().reduce(UInt64(0)) { $0 | UInt64($1.element) << (8 * UInt64($1.offset)) }
    }

    mutating func string(limit: UInt64) throws -> String {
      let length = try u64()
      guard length <= limit else { throw Failure(reason: "string too long") }
      let data = try read(Int(length))
      guard let text = String(data: data, encoding: .utf8) else { throw Failure(reason: "string is not utf-8") }
      return text
    }
  }
}

File: android/namu-core/src/main/kotlin/org/namuai/offline/core/gguf/GgufCheck.kt
package org.namuai.offline.core.gguf

import java.io.BufferedInputStream
import java.io.EOFException
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream

class GgufFormatException(message: String) : Exception(message)

sealed class GgufVerdict {
    object Ok : GgufVerdict()

    /** Structural bound violation → FILE_DAMAGED. */
    class Damaged(val reason: String) : GgufVerdict()

    /** `general.architecture` differs from the signed descriptor → MODEL_INCOMPATIBLE. */
    class Incompatible(val found: String) : GgufVerdict()
}

/** Seam so tests can prove the structural check never runs on unverified bytes (T08, DL-010). */
interface GgufInspector {
    fun check(file: File, expectedArchitecture: String): GgufVerdict
}

/**
 * Bounded structural GGUF header reader (contract §6.4 step 3), port of
 * model-release/dev/gguf-header.mjs. It walks only the metadata key/value
 * section inside a 64 MiB window, never touches tensor data and never calls
 * the native inference parser. Callers MUST have verified the SHA-256 first.
 */
object GgufCheck : GgufInspector {
    const val MAX_KV = 4096L
    const val MAX_TENSORS = 65_536L
    const val MAX_KEY_BYTES = 1 shl 20
    const val MAX_ARCHITECTURE_BYTES = 256
    const val HEADER_WINDOW = 64L * 1024 * 1024
    private const val MAX_SAFE_INTEGER = 9_007_199_254_740_991L

    private const val TYPE_STRING = 8
    private const val TYPE_ARRAY = 9

    private fun scalarSize(type: Long): Int = when (type) {
        0L, 1L, 7L -> 1
        2L, 3L -> 2
        4L, 5L, 6L -> 4
        10L, 11L, 12L -> 8
        else -> -1
    }

    override fun check(file: File, expectedArchitecture: String): GgufVerdict {
        val found = try {
            readArchitecture(file)
        } catch (e: GgufFormatException) {
            return GgufVerdict.Damaged(e.message ?: "format")
        } catch (e: IOException) {
            return GgufVerdict.Damaged("io")
        }
        return if (found == expectedArchitecture) GgufVerdict.Ok else GgufVerdict.Incompatible(found)
    }

    @Throws(GgufFormatException::class, IOException::class)
    fun readArchitecture(file: File): String {
        val window = minOf(file.length(), HEADER_WINDOW)
        FileInputStream(file).use { raw ->
            return Reader(BufferedInputStream(raw, 64 * 1024), window).readArchitecture()
        }
    }

    private class Reader(private val input: InputStream, private val window: Long) {
        private var offset = 0L
        private val scratch = ByteArray(8)

        private fun need(n: Long) {
            if (n < 0 || offset + n > window) throw GgufFormatException("metadata exceeds header window")
        }

        private fun readFully(target: ByteArray, length: Int) {
            need(length.toLong())
            var done = 0
            while (done < length) {
                val r = input.read(target, done, length - done)
                if (r < 0) throw EOFException()
                done += r
            }
            offset += length
        }

        private fun skipFully(n: Long) {
            need(n)
            var left = n
            while (left > 0) {
                val skipped = input.skip(left)
                if (skipped > 0) {
                    left -= skipped
                } else {
                    if (input.read() < 0) throw EOFException()
                    left -= 1
                }
            }
            offset += n
        }

        private fun u32(): Long {
            readFully(scratch, 4)
            return (scratch[0].toLong() and 0xff) or
                ((scratch[1].toLong() and 0xff) shl 8) or
                ((scratch[2].toLong() and 0xff) shl 16) or
                ((scratch[3].toLong() and 0xff) shl 24)
        }

        private fun u64(): Long {
            readFully(scratch, 8)
            var v = 0L
            for (i in 7 downTo 0) v = (v shl 8) or (scratch[i].toLong() and 0xff)
            if (v < 0 || v > MAX_SAFE_INTEGER) throw GgufFormatException("length out of range")
            return v
        }

        private fun str(limit: Int): String {
            val len = u64()
            if (len > limit) throw GgufFormatException("string too long")
            need(len)
            val bytes = ByteArray(len.toInt())
            readFully(bytes, bytes.size)
            return String(bytes, Charsets.UTF_8)
        }

        private fun skipValue(type: Long, depth: Int) {
            val scalar = scalarSize(type)
            when {
                scalar > 0 -> skipFully(scalar.toLong())
                type == TYPE_STRING.toLong() -> skipFully(u64())
                type == TYPE_ARRAY.toLong() -> {
                    if (depth > 1) throw GgufFormatException("nested arrays not allowed")
                    val inner = u32()
                    val count = u64()
                    val innerScalar = scalarSize(inner)
                    if (innerScalar > 0) {
                        // count <= 2^53 and size <= 8, so the product cannot overflow a Long.
                        skipFully(innerScalar.toLong() * count)
                    } else {
                        var i = 0L
                        // Every element consumes >= 8 bytes, so the window bounds this loop.
                        while (i < count) {
                            skipValue(inner, depth + 1)
                            i++
                        }
                    }
                }
                else -> throw GgufFormatException("unknown value type")
            }
        }

        fun readArchitecture(): String {
            try {
                val magic = ByteArray(4)
                readFully(magic, 4)
                if (magic[0] != 'G'.code.toByte() || magic[1] != 'G'.code.toByte() ||
                    magic[2] != 'U'.code.toByte() || magic[3] != 'F'.code.toByte()
                ) {
                    throw GgufFormatException("bad magic")
                }
                val version = u32()
                if (version != 2L && version != 3L) throw GgufFormatException("unsupported version")
                val tensors = u64()
                val kvCount = u64()
                if (tensors > MAX_TENSORS || kvCount > MAX_KV) throw GgufFormatException("counts out of bounds")
                var i = 0L
                while (i < kvCount) {
                    val key = str(MAX_KEY_BYTES)
                    val type = u32()
                    if (key == "general.architecture") {
                        if (type != TYPE_STRING.toLong()) throw GgufFormatException("architecture is not a string")
                        return str(MAX_ARCHITECTURE_BYTES)
                    }
                    skipValue(type, 0)
                    i++
                }
                throw GgufFormatException("general.architecture missing")
            } catch (e: EOFException) {
                throw GgufFormatException("metadata exceeds header window")
            }
        }
    }
}
